perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/Vsup.java

```diff
@@ -40,55 +44,88 @@
+	public void canBeApproved(PerunSession session, Application app) throws PerunException {
 
-		if (application.getUser() == null) {
+		List<ApplicationFormItemData> data = registrar.getApplicationDataById(session, app.getId());
+
+		if (app.getUser() == null) {
+
+			for (ApplicationFormItemData item : data) {
+				if (item.getFormItem() != null &&
+					Objects.equals(AttributesManager.NS_USER_ATTR_DEF + ":birthNumber", item.getFormItem().getPerunDestinationAttribute())) {
+
+					// if application contains birth number, try to map to existing user
+					String rc = item.getValue();
+					if (rc != null && !rc.isEmpty()) {
+
+						try {
+							User user = ((PerunBl) session.getPerun()).getUsersManagerBl().getUserByExtSourceNameAndExtLogin(session, "RC", rc);
+							throw new CantBeApprovedException("Application has the same birth number " + rc + " as user " + user.getDisplayName() + " with id " + user.getId() +
+								" that is already in Perun and thus would be merged with him.", null, null, null, true);
+						} catch (CantBeApprovedException ex) {
+							throw ex;
+						} catch (Exception ex) {
+							log.warn("Couldn't find or set user to application {} by RC: {}", app, ex);
+						}
+					}
+					break;
+				}
+			}
+		}
+	}
+
+	@Override
+	public Application beforeApprove(PerunSession session, Application app) throws RegistrarException, PrivilegeException {
+
+		List<ApplicationFormItemData> data = registrar.getApplicationDataById(session, app.getId());
+
+		if (app.getUser() == null) {
 
 			for (ApplicationFormItemData item : data) {
 				if (item.getFormItem() != null &&
-						Objects.equals(AttributesManager.NS_USER_ATTR_DEF+":birthNumber", item.getFormItem().getPerunDestinationAttribute())) {
+					Objects.equals(AttributesManager.NS_USER_ATTR_DEF + ":birthNumber", item.getFormItem().getPerunDestinationAttribute())) {
 
 					// if application contains birth number, try to map to existing user
 					String rc = item.getValue();
 					if (rc != null && !rc.isEmpty()) {
 
 						try {
 							User user = ((PerunBl) session.getPerun()).getUsersManagerBl().getUserByExtSourceNameAndExtLogin(session, "RC", rc);
-							application.setUser(user);
-							registrar.updateApplicationUser(session, application);
-							log.debug("Existing user found by RC for {}", application);
+							app.setUser(user);
+							registrar.updateApplicationUser(session, app);
+							log.debug("Existing user found by RC for {}", app);
 						} catch (Exception ex) {
-							log.warn("Couldn't find or set user to application {} by RC: {}", application, ex);
+							log.warn("Couldn't find or set user to application {} by RC: {}", app, ex);
 						}
 
 						// associate existing user with the identity used on registration form
-						if (application.getUser() != null) {
+						if (app.getUser() != null) {
```

Comment:
I realized, that this won't handle the cases, when found user is already a member, since after this method, normal approval process takes place. So here, for case when user is not null, please add check, whether user is already a member of VO, eg.:
```java
try {
	Member member = ((PerunBl) session.getPerun()).getMembersManagerBl().getMemberByUser(session, app.getVo(), app.getUser());
	// user is already a member, switch application type
	if (Application.AppType.INITIAL.equals(app.getType())) {
		app.setType(Application.AppType.EXTENSION);
		registrar.updateApplicationType(session, app);
		log.debug("Updating application type to EXTENSION since we matched user which is VO member!");
	}
} catch (MemberNotExistsException e) {
	// OK state
}
```
You must also implement new method in the RegistrarManager to update application type in a DB. It must be public, but there is no need to add it to the RPC, so we don't need authorization in it yet.

Reply:
Okay, thank you for comment, I've added this check and also the mentioned method.